samples/client.c

```diff
@@ -62,6 +62,8 @@ static char const *irq_to_str[] = {
     [VFU_DEV_REQ_IRQ] = "REQ"
 };
 
+static pthread_mutex_t mutex = PTHREAD_MUTEX_INITIALIZER;
```

Comment:
We don't, I think it's a remnant of some debugging from last week that I accidentally forgot to remove - I'll get rid of it!

Reply:
The fake guest is calling `access_region` while we do the migration so we have to ensure that the responses don't get mixed up.

Reply:
Why is this a problem with migration v2 and not v1?

Reply:
I think it might've been an issue with v1 too - every so often running the sample from the master branch you randomly get a CRC mismatch (but it doesn't seem to completely fall over like with v2 without the mutex).

Reply:
This doesn't seem to fix the problem - I have it now setting the `no_reply` flag on writes to BAR1 in the master version but the CRC mismatch still occurs...

Reply:
Wait I've figured it out, it's not a concurrency bug, it's that in `fake_guest` we calculate the CRC with the previous CRC as the initialisation bytes (incorrectly), so if the fake guest does more than one iteration the CRC will not match.

Reply:
That's a different problem then, let's skip fixing master for now.

Back to v2, if you set `no_reply` you shouldn't be getting mixed responses, so there's shouldn't be a need for additional locking to prevent this. CRC mismatch errors should be irrelevant here.

Regarding CRC mismatch errors, here's what I had in mind when implementing the client/server sample:

- the fake guest constantly writes to BAR1 and store the CRC of what it wrote
- the main thread sets migration state to PRE_COPY and does a few, pointless save iterations. It doesn't care about CRC. IIUC the fake guest does at least one iteration.
- the main thread stops the fake guest, sets migration state to SAVING and does a few saving iterations to read device state. Ideally it should do at least two such "iterations", where it should read BAR1 in at least two steps to be more realistic (I realize that iterations is the wrong term, it should be steps).
- the main thread forks and creates the destination servers, sets device state to RESUMING, applies device state in at lest two steps (again, misusing term iterations), sets device state to running, validates CRC but does not restart the fake guest.

So I _believe_ there shouldn't be any CRC mismatch errors, obviously if you still get them then there's something wrong. First of all, could you add this description somewhere? Second, could you fix the misuse of iterations, e.g. replace with steps? Third, can you add some assertions where we execute at least two save/resume steps?